bakery/mgostorage/rootkey.go

```diff
@@ -0,0 +1,348 @@
+	}
+	return &rootKeyStorage{
+		keys:   s,
+		coll:   c,
+		policy: policy,
+	}
+}
+
+var indexes = []mgo.Index{{
+	Key: []string{"-created"},
+}, {
+	Key:         []string{"expires"},
+	ExpireAfter: time.Second,
+}}
+
+// EnsureIndex ensures that the required indexes exist on the
+// collection that will be used for root key storage.
+// This should be called at least once before using NewStorage.
+func (s *RootKeys) EnsureIndex(c *mgo.Collection) error {
+	for _, idx := range indexes {
+		if err := c.EnsureIndex(idx); err != nil {
+			return errgo.Notef(err, "cannot ensure index for %q on %q", idx.Key, c.Name)
+		}
+	}
+	return nil
+}
+
+// get gets the root key for the given id, trying the cache first and
+// falling back to calling fallback if it's not found there.
+//
+// If the key does not exist or has expired, it returns
+// bakery.ErrNotFound.
+//
+// Called with s.mu locked.
+func (s *RootKeys) get(id string, fallback func(id string) (rootKey, error)) (rootKey, error) {
+	key, cached, err := s.get0(id, fallback)
+	if err != nil && err != bakery.ErrNotFound {
+		return rootKey{}, errgo.Mask(err)
+	}
+	if err == nil && timeNow().After(key.Expires) {
+		key = rootKey{}
+		err = bakery.ErrNotFound
+	}
+	if !cached {
+		s.addCache(id, key)
+	}
+	return key, err
+}
+
+// get0 is the inner version of RootKeys.get. It returns an item and reports
+// whether it was found in the cache, but doesn't check whether the
+// item has expired or move the returned item to s.cache.
+func (s *RootKeys) get0(id string, fallback func(id string) (rootKey, error)) (key rootKey, inCache bool, err error) {
+	if k, ok := s.cache[id]; ok {
+		if !k.isValid() {
+			return rootKey{}, true, bakery.ErrNotFound
+		}
+		return k, true, nil
+	}
+	if k, ok := s.oldCache[id]; ok {
+		if !k.isValid() {
+			return rootKey{}, false, bakery.ErrNotFound
+		}
+		return k, false, nil
+	}
+	logger.Infof("cache miss for %q", id)
+	k, err := fallback(id)
+	return k, false, err
+}
+
+// addCache adds the given key to the cache.
+// Called with s.mu locked.
+func (s *RootKeys) addCache(id string, k rootKey) {
+	if len(s.cache) >= s.maxCacheSize {
+		s.oldCache = s.cache
+		s.cache = make(map[string]rootKey)
+	}
+	s.cache[id] = k
+}
+
+// setCurrent sets the current key for the given storage policy.
+// Called with s.mu locked.
+func (s *RootKeys) setCurrent(policy Policy, key rootKey) {
+	if len(s.current) > maxPolicyCache {
+		// Sanity check to avoid possibly memory leak:
+		// if some client is using arbitrarily many storage
+		// policies, we don't want s.keys.current to endlessly
+		// expand, so just kill the cache if it grows too big.
+		// This will result in worse performance but it shouldn't
+		// happen in practice and it's better than using endless
+		// space.
+		s.current = make(map[Policy]rootKey)
+	}
+	s.current[policy] = key
+}
+
+type rootKeyStorage struct {
+	keys   *RootKeys
+	policy Policy
+	coll   *mgo.Collection
```

Comment:
I feel like we shouldn't be holding on to a collection like this.  We can take a collection name as string and accept a session that we use for connecting. Otherwise we'll need to create a new rootKeyStorage when our connection to mongo drops